File: Problem1_CombinationSum.py
# Time Complexity: O(2^n), where n is length of the input array
# Space Complexity: O(n)
# Did this code successfully run on Leetcode: Yes

# Solution:

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        if not candidates or len(candidates) == 0:
            return []

        self.result = []
        self.dfs(candidates, target, 0, [])
        return self.result

    def dfs(self, candidates: List[int], target: int, index: int, path: List[int]) -> None:
        # base
        # If target is reached, add the path to the result
        if target == 0:
            self.result.append(list(path))
            return

        # Return when target becomes negative or index is out of bounds
        if target < 0 or index == len(candidates):
            return

        # logic
        # don't choose
        self.dfs(candidates, target, index + 1, path)
        # choose
        path.append(candidates[index])
        self.dfs(candidates, target - candidates[index], index, path)
        # backtrack
        path.pop()

File: Problem1_CombinationSum.java
// Time Complexity : exponential
// Space Complexity : O(n) recursive stack
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    List<List<Integer>> result;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        result=new ArrayList<List<Integer>>();
        
        if(candidates==null || candidates.length==0)
            return result;
        
        helper(candidates,target,0,new ArrayList<Integer>());
        
        return result;
    }
    /* recursion */
    /*
    private void helper(int[] candidates,int target,int index,List<Integer> path)
    {
        if(target==0)
        {
            result.add(new ArrayList<>(path));
            return;
        }
            
        
        if(target<0 || index==candidates.length)
            return;
        
        //dont choose
         helper(candidates,target,index+1,path);
        //choose
        path.add(candidates[index]);
         helper(candidates,target-candidates[index],index,path);
        path.remove(path.size()-1);
    }
    */
    
    //backtra cking
     private void helper(int[] candidates,int target,int index,List<Integer> path)
    {
        if(target==0)
        {
            result.add(new ArrayList<>(path));
            return;
        }
            
        
        if(target<0)
            return;
        
        for(int i=index;i<candidates.length;i++)
        {
             path.add(candidates[i]);
             helper(candidates,target-candidates[i],i,path);
             path.remove(path.size()-1);
        }
       
    }
}